### PR TITLE
layers: Add sampler unnormalized coords checks

### LINKED ISSUE #156: Implement VkSamplerCreateInfo-unnormalizedCoordinates-*
These VUIDs aren't implemented according to the database, and would have saved us a fair amount of time (and stress, due to deadlines) recently.

* If unnormalizedCoordinates is VK_TRUE, minFilter and magFilter must be equal
* If unnormalizedCoordinates is VK_TRUE, mipmapMode must be VK_SAMPLER_MIPMAP_MODE_NEAREST
* If unnormalizedCoordinates is VK_TRUE, minLod and maxLod must be zero
* If unnormalizedCoordinates is VK_TRUE, addressModeU and addressModeV must each be either VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE or VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER
* If unnormalizedCoordinates is VK_TRUE, anisotropyEnable must be VK_FALSE
* If unnormalizedCoordinates is VK_TRUE, compareEnable must be VK_FALSE

Since this looks really easy, I may try to use it as a Hello Validation Layers project for myself this weekend.

### PR DESCRIPTION
**Pull request by critsec**
Adds checks for VUIDs:
* VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01072
* VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01073
* VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01074
* VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01075
* VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01077

and tests them with:
* VkLayerTest.UnnormalizedCoordinatesEnabled

Fixes #156

### CODE REVIEW

**Comment by critsec** [comment]
CI failures seem to be inability to find Vulkan-Headers? My change shouldn't affect that, but previous changes aren't failing. Not sure what's going on there.

**Comment by tobine** [comment]
The CI errors are caused by recent update to Vulkan-Loaders repo by @lenny-lunarg . I filed [26](https://github.com/KhronosGroup/Vulkan-Loader/issues/26) on this issue.

The repo split has currently caused the Layers CI to be non-deterministic as repo dependencies are being pulled from HEAD instead of known-good revs. I'll file a separate issue for this if there isn't one already.

@lenny-lunarg the CI config for ValidationLayers needs to be updated.

File: layers/parameter_validation_utils.cpp [hunk 1]
```diff
@@ -2037,14 +2038,53 @@ bool pv_vkCreateSampler(VkDevice device, const VkSamplerCreateInfo *pCreateInfo,
...
+                                "vkCreateSampler(): when pCreateInfo->unnormalizedCoordinates is VK_TRUE, "
+                                "pCreateInfo->minFilter (%d) and pCreateInfo->magFilter (%d) must be equal.",
+                                pCreateInfo->minFilter, pCreateInfo->magFilter);
+            }
+            if (pCreateInfo->mipmapMode != VK_SAMPLER_MIPMAP_MODE_NEAREST) {
+                skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
+                                "VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01073",
+                                "vkCreateSampler(): when pCreateInfo->unnormalizedCoordinates is VK_TRUE, "
+                                "pCreateInfo->mipmapMode (%d) must be VK_SAMPLER_MIPMAP_MODE_NEAREST.",
+                                pCreateInfo->mipmapMode);
```
Comment (tobine) [inline layers/parameter_validation_utils.cpp]: Can use helper function string_VkSamplerMipmapMode() here to print mipmapMode as string.

File: layers/parameter_validation_utils.cpp [hunk 2]
```diff
@@ -2037,14 +2038,53 @@ bool pv_vkCreateSampler(VkDevice device, const VkSamplerCreateInfo *pCreateInfo,
...
 
-            // Anistropy and unnormalized coordinates cannot be enabled simultaneously
-            if (pCreateInfo->unnormalizedCoordinates == VK_TRUE) {
+        if (pCreateInfo->unnormalizedCoordinates == VK_TRUE) {
+            if (pCreateInfo->minFilter != pCreateInfo->magFilter) {
+                skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
+                                "VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01072",
+                                "vkCreateSampler(): when pCreateInfo->unnormalizedCoordinates is VK_TRUE, "
+                                "pCreateInfo->minFilter (%d) and pCreateInfo->magFilter (%d) must be equal.",
+                                pCreateInfo->minFilter, pCreateInfo->magFilter);
```
Comment (tobine) [inline layers/parameter_validation_utils.cpp]: Can use helper function string_VkFilter() to print min/magFilter values as strings.

File: layers/parameter_validation_utils.cpp [hunk 3]
```diff
@@ -2037,14 +2038,53 @@ bool pv_vkCreateSampler(VkDevice device, const VkSamplerCreateInfo *pCreateInfo,
...
+            if ((pCreateInfo->addressModeU != VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE &&
+                 pCreateInfo->addressModeU != VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER) ||
+                (pCreateInfo->addressModeV != VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE &&
+                 pCreateInfo->addressModeV != VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER)) {
+                skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
+                                "VUID-VkSamplerCreateInfo-unnormalizedCoordinates-01075",
+                                "vkCreateSampler(): when pCreateInfo->unnormalizedCoordinates is VK_TRUE, "
+                                "pCreateInfo->addressModeU (%d) and pCreateInfo->addressModeV (%d) must both be "
+                                "VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE or VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER.",
+                                pCreateInfo->addressModeU, pCreateInfo->addressModeV);
```
[the remark this answers is not in the document]
Comment (critsec) [reply layers/parameter_validation_utils.cpp]: Thanks! I looked for something like that but didn't find it; not sure why as it's obvious in hindsight and my greps should have found it. Oh well, fixed now.

File: layers/parameter_validation_utils.cpp (hunk 1 above)
Comment (critsec) [reply layers/parameter_validation_utils.cpp]: Fixed

File: layers/parameter_validation_utils.cpp (hunk 2 above)
Comment (critsec) [reply layers/parameter_validation_utils.cpp]: Fixed

**Comment by karl-lunarg** [comment]
If you rebase from master now to pick up fixes for #162, then your CI should work.

**Comment by critsec** [comment]
Rebased and CI is passing.

**Pull request merged**